<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>Throwables.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">JaCoCo Cassandara Coverage Report</a> &gt; <a href="index.source.html" class="el_package">org.apache.cassandra.utils</a> &gt; <span class="el_source">Throwables.java</span></div><h1>Throwables.java</h1><pre class="source lang-java linenums">/*
* Licensed to the Apache Software Foundation (ASF) under one
* or more contributor license agreements.  See the NOTICE file
* distributed with this work for additional information
* regarding copyright ownership.  The ASF licenses this file
* to you under the Apache License, Version 2.0 (the
* &quot;License&quot;); you may not use this file except in compliance
* with the License.  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing,
* software distributed under the License is distributed on an
* &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied.  See the License for the
* specific language governing permissions and limitations
* under the License.
*/
package org.apache.cassandra.utils;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.Iterator;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ExecutionException;
import java.util.function.Predicate;
import java.util.stream.Stream;
import javax.annotation.Nonnull;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.Iterables;

import org.apache.cassandra.io.FSReadError;
import org.apache.cassandra.io.FSWriteError;
import org.apache.cassandra.io.util.File;
import org.apache.cassandra.utils.concurrent.UncheckedInterruptedException;

<span class="nc" id="L42">public final class Throwables</span>
{
<span class="fc" id="L44">    public enum FileOpType { READ, WRITE }</span>

    public interface DiscreteAction&lt;E extends Exception&gt;
    {
        void perform() throws E;
    }

    public static boolean isCausedBy(Throwable t, Predicate&lt;Throwable&gt; cause)
    {
<span class="pc bpc" id="L53" title="1 of 6 branches missed.">        return cause.test(t) || (t.getCause() != null &amp;&amp; cause.test(t.getCause()));</span>
    }

    public static boolean anyCauseMatches(Throwable t, Predicate&lt;Throwable&gt; cause)
    {
        do
        {
<span class="fc bfc" id="L60" title="All 2 branches covered.">            if (cause.test(t))</span>
<span class="fc" id="L61">                return true;</span>
<span class="fc bfc" id="L62" title="All 2 branches covered.">        } while ((t = t.getCause()) != null);</span>
<span class="fc" id="L63">        return false;</span>
    }

    public static &lt;T extends Throwable&gt; T merge(T existingFail, T newFail)
    {
<span class="pc bpc" id="L68" title="1 of 2 branches missed.">        if (existingFail == null)</span>
<span class="fc" id="L69">            return newFail;</span>
<span class="nc bnc" id="L70" title="All 2 branches missed.">        if (newFail == null)</span>
<span class="nc" id="L71">            return existingFail;</span>
<span class="nc" id="L72">        existingFail.addSuppressed(newFail);</span>
<span class="nc" id="L73">        return existingFail;</span>
    }

    public static void maybeFail(Throwable fail)
    {
<span class="fc bfc" id="L78" title="All 2 branches covered.">        if (failIfCanCast(fail, null))</span>
<span class="fc" id="L79">            throw new RuntimeException(fail);</span>
<span class="fc" id="L80">    }</span>

    public static &lt;T extends Throwable&gt; void maybeFail(Throwable fail, Class&lt;T&gt; checked) throws T
    {
<span class="pc bpc" id="L84" title="1 of 2 branches missed.">        if (failIfCanCast(fail, checked))</span>
<span class="nc" id="L85">            throw new RuntimeException(fail);</span>
<span class="fc" id="L86">    }</span>

    public static &lt;T extends Throwable&gt; boolean failIfCanCast(Throwable fail, Class&lt;T&gt; checked) throws T
    {
<span class="fc bfc" id="L90" title="All 2 branches covered.">        if (fail == null)</span>
<span class="fc" id="L91">            return false;</span>

<span class="pc bpc" id="L93" title="1 of 2 branches missed.">        if (fail instanceof Error)</span>
<span class="nc" id="L94">            throw (Error) fail;</span>

<span class="fc bfc" id="L96" title="All 2 branches covered.">        if (fail instanceof RuntimeException)</span>
<span class="fc" id="L97">            throw (RuntimeException) fail;</span>

<span class="pc bpc" id="L99" title="1 of 2 branches missed.">        if (fail instanceof InterruptedException)</span>
<span class="nc" id="L100">            throw new UncheckedInterruptedException((InterruptedException) fail);</span>

<span class="pc bpc" id="L102" title="3 of 4 branches missed.">        if (checked != null &amp;&amp; checked.isInstance(fail))</span>
<span class="nc" id="L103">            throw checked.cast(fail);</span>

<span class="fc" id="L105">        return true;</span>
    }

    @SafeVarargs
    public static &lt;E extends Exception&gt; void maybeFail(DiscreteAction&lt;? extends E&gt; ... actions)
    {
<span class="fc" id="L111">        maybeFail(Throwables.perform(null, Stream.of(actions)));</span>
<span class="fc" id="L112">    }</span>

    @SafeVarargs
    public static &lt;E extends Exception&gt; void perform(DiscreteAction&lt;? extends E&gt; ... actions) throws E
    {
<span class="nc" id="L117">        Throwables.&lt;E&gt;perform(Stream.of(actions));</span>
<span class="nc" id="L118">    }</span>

    public static &lt;E extends Exception&gt; void perform(Stream&lt;? extends DiscreteAction&lt;? extends E&gt;&gt; stream, DiscreteAction&lt;? extends E&gt; ... extra) throws E
    {
<span class="nc" id="L122">        perform(Stream.concat(stream, Stream.of(extra)));</span>
<span class="nc" id="L123">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    public static &lt;E extends Exception&gt; void perform(Stream&lt;DiscreteAction&lt;? extends E&gt;&gt; actions) throws E
    {
<span class="nc" id="L128">        Throwable fail = perform(null, actions);</span>
<span class="nc bnc" id="L129" title="All 2 branches missed.">        if (failIfCanCast(fail, null))</span>
<span class="nc" id="L130">            throw (E) fail;</span>
<span class="nc" id="L131">    }</span>

    public static Throwable perform(Throwable accumulate, DiscreteAction&lt;?&gt; ... actions)
    {
<span class="fc" id="L135">        return perform(accumulate, Arrays.stream(actions));</span>
    }

    public static Throwable perform(Throwable accumulate, Stream&lt;? extends DiscreteAction&lt;?&gt;&gt; actions)
    {
<span class="fc" id="L140">        return perform(accumulate, actions.iterator());</span>
    }

    public static Throwable perform(Throwable accumulate, Iterator&lt;? extends DiscreteAction&lt;?&gt;&gt; actions)
    {
<span class="fc bfc" id="L145" title="All 2 branches covered.">        while (actions.hasNext())</span>
        {
<span class="fc" id="L147">            DiscreteAction&lt;?&gt; action = actions.next();</span>
            try
            {
<span class="fc" id="L150">                action.perform();</span>
            }
<span class="nc" id="L152">            catch (Throwable t)</span>
            {
<span class="nc" id="L154">                accumulate = merge(accumulate, t);</span>
<span class="fc" id="L155">            }</span>
<span class="fc" id="L156">        }</span>
<span class="fc" id="L157">        return accumulate;</span>
    }

    @SafeVarargs
    public static void perform(File against, FileOpType opType, DiscreteAction&lt;? extends IOException&gt; ... actions)
    {
<span class="fc" id="L163">        perform(against.path(), opType, actions);</span>
<span class="fc" id="L164">    }</span>

    @SafeVarargs
    public static void perform(String filePath, FileOpType opType, DiscreteAction&lt;? extends IOException&gt; ... actions)
    {
<span class="fc" id="L169">        maybeFail(perform(null, filePath, opType, actions));</span>
<span class="fc" id="L170">    }</span>

    @SafeVarargs
    public static Throwable perform(Throwable accumulate, String filePath, FileOpType opType, DiscreteAction&lt;? extends IOException&gt; ... actions)
    {
<span class="fc" id="L175">        return perform(accumulate, filePath, opType, Arrays.stream(actions));</span>
    }

    public static Throwable perform(Throwable accumulate, String filePath, FileOpType opType, Stream&lt;DiscreteAction&lt;? extends IOException&gt;&gt; actions)
    {
<span class="fc" id="L180">        return perform(accumulate, actions.map((action) -&gt; () -&gt;</span>
        {
            try
            {
<span class="fc" id="L184">                action.perform();</span>
            }
<span class="nc" id="L186">            catch (IOException e)</span>
            {
<span class="nc bnc" id="L188" title="All 2 branches missed.">                throw (opType == FileOpType.WRITE) ? new FSWriteError(e, filePath) : new FSReadError(e, filePath);</span>
<span class="fc" id="L189">            }</span>
<span class="fc" id="L190">        }));</span>
    }

    /**
     * @see {@link #closeAndAddSuppressed(Throwable, Iterable)}
     */
    public static void closeAndAddSuppressed(@Nonnull Throwable t, AutoCloseable... closeables)
    {
<span class="nc" id="L198">        closeAndAddSuppressed(t, Arrays.asList(closeables));</span>
<span class="nc" id="L199">    }</span>

    /**
     * Do what {@link #closeAndAddSuppressed(Throwable, Iterable)} does, additionally filtering out all null closables.
     */
    public static void closeNonNullAndAddSuppressed(@Nonnull Throwable t, AutoCloseable... closeables)
    {
<span class="nc" id="L206">        closeAndAddSuppressed(t, Iterables.filter(Arrays.asList(closeables), Objects::nonNull));</span>
<span class="nc" id="L207">    }</span>

    /**
     * Closes all closables in the provided collections and accumulates the possible exceptions thrown when closing.
     *
     * @param accumulate non-null exception to accumulate errors thrown when closing the provided resources
     * @param closeables closeables to be closed
     */
    public static void closeAndAddSuppressed(@Nonnull Throwable accumulate, Iterable&lt;AutoCloseable&gt; closeables)
    {
<span class="nc" id="L217">        Preconditions.checkNotNull(accumulate);</span>
<span class="nc bnc" id="L218" title="All 2 branches missed.">        for (AutoCloseable closeable : closeables)</span>
        {
            try
            {
<span class="nc" id="L222">                closeable.close();</span>
            }
<span class="nc" id="L224">            catch (Throwable ex)</span>
            {
<span class="nc" id="L226">                accumulate.addSuppressed(ex);</span>
<span class="nc" id="L227">            }</span>
<span class="nc" id="L228">        }</span>
<span class="nc" id="L229">    }</span>

    /**
     * See {@link #close(Throwable, Iterable)}
     */
    public static Throwable close(Throwable accumulate, AutoCloseable ... closeables)
    {
<span class="fc" id="L236">        return close(accumulate, Arrays.asList(closeables));</span>
    }

    /**
     * Closes all the resources in the provided collections and accumulates the possible exceptions thrown when closing.
     *
     * @param accumulate the initial value for the exception accumulator, can be {@code null}
     * @param closeables closeables to be closed
     * @return {@code null}, {@param accumulate} or the first exception thrown when closing the provided resources
     */
    public static Throwable close(Throwable accumulate, Iterable&lt;? extends AutoCloseable&gt; closeables)
    {
<span class="fc bfc" id="L248" title="All 2 branches covered.">        for (AutoCloseable closeable : closeables)</span>
        {
            try
            {
<span class="fc" id="L252">                closeable.close();</span>
            }
<span class="nc" id="L254">            catch (Throwable t)</span>
            {
<span class="nc" id="L256">                accumulate = merge(accumulate, t);</span>
<span class="fc" id="L257">            }</span>
<span class="fc" id="L258">        }</span>
<span class="fc" id="L259">        return accumulate;</span>
    }

    public static Optional&lt;IOException&gt; extractIOExceptionCause(Throwable t)
    {
<span class="nc bnc" id="L264" title="All 2 branches missed.">        if (t instanceof IOException)</span>
<span class="nc" id="L265">            return Optional.of((IOException) t);</span>
<span class="nc" id="L266">        Throwable cause = t;</span>
<span class="nc bnc" id="L267" title="All 2 branches missed.">        while ((cause = cause.getCause()) != null)</span>
        {
<span class="nc bnc" id="L269" title="All 2 branches missed.">            if (cause instanceof IOException)</span>
<span class="nc" id="L270">                return Optional.of((IOException) cause);</span>
        }
<span class="nc" id="L272">        return Optional.empty();</span>
    }

    /**
     * If the provided throwable is a &quot;wrapping&quot; exception (see below), return the cause of that throwable, otherwise
     * return its argument untouched.
     * &lt;p&gt;
     * We call a &quot;wrapping&quot; exception in the context of that method an exception whose only purpose is to wrap another
     * exception, and currently this method recognize only 2 exception as &quot;wrapping&quot; ones: {@link ExecutionException}
     * and {@link CompletionException}.
     */
    public static Throwable unwrapped(Throwable t)
    {
<span class="nc" id="L285">        Throwable unwrapped = t;</span>
<span class="nc bnc" id="L286" title="All 6 branches missed.">        while (unwrapped instanceof CompletionException ||</span>
               unwrapped instanceof ExecutionException ||
               unwrapped instanceof InvocationTargetException)
<span class="nc" id="L289">            unwrapped = unwrapped.getCause();</span>

        // I don't think it make sense for those 2 exception classes to ever be used with null causes, but no point
        // in failing here if this happen. We still wrap the original exception if that happen so we get a sign
        // that the assumption of this method is wrong.
<span class="nc bnc" id="L294" title="All 2 branches missed.">        return unwrapped == null</span>
<span class="nc" id="L295">               ? new RuntimeException(&quot;Got wrapping exception not wrapping anything&quot;, t)</span>
<span class="nc" id="L296">               : unwrapped;</span>
    }

    /**
     * If the provided exception is unchecked, return it directly, otherwise wrap it into a {@link RuntimeException}
     * to make it unchecked.
     */
    public static RuntimeException unchecked(Throwable t)
    {
<span class="fc bfc" id="L305" title="All 2 branches covered.">        return t instanceof RuntimeException ? (RuntimeException)t :</span>
<span class="pc bpc" id="L306" title="1 of 2 branches missed.">               t instanceof InterruptedException</span>
<span class="nc" id="L307">               ? new UncheckedInterruptedException((InterruptedException) t)</span>
<span class="fc" id="L308">               : new RuntimeException(t);</span>
    }

    /**
     * throw the exception as a unchecked exception, wrapping if a checked exception, else rethroing as is.
     */
    public static RuntimeException throwAsUncheckedException(Throwable t)
    {
<span class="fc bfc" id="L316" title="All 2 branches covered.">        if (t instanceof Error)</span>
<span class="fc" id="L317">            throw (Error) t;</span>
<span class="fc" id="L318">        throw unchecked(t);</span>
    }

    /**
     * A shortcut for {@code unchecked(unwrapped(t))}. This is called &quot;cleaned&quot; because this basically removes the annoying
     * cruft surrounding an exception :).
     */
    public static RuntimeException cleaned(Throwable t)
    {
<span class="nc" id="L327">        return unchecked(unwrapped(t));</span>
    }

    @VisibleForTesting
    public static void assertAnyCause(Throwable err, Class&lt;? extends Throwable&gt; cause)
    {
<span class="nc bnc" id="L333" title="All 2 branches missed.">        if (!anyCauseMatches(err, cause::isInstance))</span>
<span class="nc" id="L334">            throw new AssertionError(&quot;The exception is not caused by &quot; + cause.getName(), err);</span>
<span class="nc" id="L335">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.8.202204050719</span></div></body></html>